spotbugs/src/main/java/edu/umd/cs/findbugs/detect/PublicAttributes.java

```diff
@@ -0,0 +1,210 @@
+                return;
+            }
+
+            // Only consider attributes of self.
+            if (!field.getClassDescriptor().equals(getClassDescriptor())) {
+                return;
+            }
+
+            if (!field.isPublic()) {
+                return;
+            }
+
+            bugReporter.reportBug(new BugInstance(this, "PA_PUBLIC_ATTRIBUTES",
+                    NORMAL_PRIORITY)
+                            .addClass(this).addField(field));
+            writtenFields.add(field);
+        } else if (seen == Const.INVOKEINTERFACE ||
+                seen == Const.INVOKEVIRTUAL) {
+            XMethod xmo = getXMethodOperand();
+            if (xmo == null) {
+                return;
+            }
+
+            // Heuristics: suppose that that model has a proper name in English
+            // which roughly describes its behavior, beginning with the verb.
+            // If the verb does not hint that the method writes the object
+            // then we skip it.
+            if (!looksLikeASetter(xmo.getName())) {
+                return;
+            }
+
+            XField field = stack.getStackItem(xmo.getNumParams()).getXField();
+            // Do not report the same public field twice.
+            if (field == null || writtenFields.contains(field)) {
+                return;
+            }
+
+            // Only consider attributes of self.
+            if (!field.getClassDescriptor().equals(getClassDescriptor())) {
+                return;
+            }
+
+            if (!field.isPublic() || !field.isReferenceType()) {
+                return;
+            }
+
+            Field clsField = getClassField(field, getClassName());
+            Type type = clsField.getType();
+            // Instances of an immutable classes cannot be modified, thus
+            // the method is not a modifier, despite its name.
+            if (type instanceof ObjectType &&
+                    isImmutable(((ObjectType) type).getClassName())) {
+                return;
+            }
+
+            bugReporter.reportBug(new BugInstance(this, "PA_PUBLIC_ATTRIBUTES",
+                    NORMAL_PRIORITY)
+                            .addClass(this).addField(field));
+            writtenFields.add(field);
+        }
+    }
+
+    private static boolean isConstructorLikeMethod(String methodName) {
+        return Const.CONSTRUCTOR_NAME.equals(methodName) ||
+                Const.STATIC_INITIALIZER_NAME.equals(methodName) ||
+                "clone".equals(methodName) ||
+                "init".equals(methodName) ||
+                "initialize".equals(methodName) ||
+                "dispose".equals(methodName) ||
+                "finalize".equals(methodName) ||
+                "this".equals(methodName) ||
+                "_jspInit".equals(methodName) ||
+                "_jspDestroy".equals(methodName);
+    }
+
+    private static Field getClassField(XField field, String className) {
```

Comment:
For maintainability, I suggest to:

1. rename `className` to `dottedClassName`, or
2. annotate `className` with `@DottedClassName`

In this repo, we have two kinds of class names: `dotted.ClassName` and `slashed/ClassName`.
It's nice to use naming or annotation to make it clear.